tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc

```diff
@@ -0,0 +1,955 @@
+// convert nodes between reduce and Subtract.
+// TODO TJ: Make this more general to any patterns that has this structure when
+// cudnn runner supports generic cudnnOpGraphs. producer
+// |   \
+// |  reduce
+// |     |
+// |  broadcast
+// |   /
+// root
+auto GetUnfusedReduceMaxSumSoftmaxPattern(
+    HloInstruction** softmax_input = nullptr) {
+  // The reduce-max part of the softmax
+  auto unfused_softmax_max_subpattern = m::Subtract(
+      m::Op(),
+      m::Broadcast(OptionalConvert(OptionalConvert(
+          m::Op()
+              .WithPredicate(IsReduceMax)
+              .WithOperand(0, OptionalConvert(m::Op(softmax_input)))))));
+  // The reduce-add part of the softmax
+  auto unfused_softmax_sum_subpattern = m::Divide(
+      m::Exp(unfused_softmax_max_subpattern),
+      m::Broadcast(OptionalConvert(OptionalConvert(
+                       m::Op()
+                           .WithOperand(0, OptionalConvert(m::Exp(
+                                               unfused_softmax_max_subpattern)))
+                           .WithPredicate(IsReduceSum)
+                           .WithOneUse())))
+          .WithOneUse());
+  return unfused_softmax_sum_subpattern;
+}
+
+// This function matches strictly reducemax-reducesum softmax patterns.
+bool IsReduceMaxSumSoftmaxFusion(const HloInstruction* instr) {
+  if (instr->called_computations().size() > 1) {
+    // Fused softmax can only have 1 called computation.
+    return false;
+  }
+  return Match(instr->called_computations()[0]->root_instruction(),
+               GetUnfusedReduceMaxSumSoftmaxPattern());
+}
+
+std::optional<double> GetConstantValue(const HloInstruction* inst) {
+  if (!IsScalar(inst)) {
+    return std::nullopt;
+  }
+  switch (inst->shape().element_type()) {
+    case F16:
+      return static_cast<float>(inst->literal().GetFirstElement<half>());
+    case BF16:
+      return static_cast<float>(inst->literal().GetFirstElement<bfloat16>());
+    case F32:
+      return inst->literal().GetFirstElement<float>();
+    case F64:
+      return inst->literal().GetFirstElement<double>();
+    default:
+      return std::nullopt;
+  }
+}
+
+double GetDropoutRateFromHlo(HloInstruction* dropout) {
+  std::optional<double> dropout_rate_inv;
+  dropout_rate_inv = GetConstantValue(dropout);
+  if (!dropout_rate_inv.has_value()) {
+    return 0.0;
+  }
+  // In dropout, inputs are divided by (1 - rate), we need to divide 1 by
+  // the constant in dropout node and substract
+  // from 1 here to get the actual dropout rate.
+  return (1.0 - (1.0 / *dropout_rate_inv));
+}
+
+HloInstruction* GetBiasFromBmm(HloInstruction* bmm) {
+  return bmm->mutable_operand(2);
+}
+
+// Give this instruction a more useful name than "custom-call.42".
+Status SetName(HloModule* module, HloInstruction* fmha) {
+  if (fmha->custom_call_target() == kCudnnfMHABmmBmmCallTarget) {
+    module->SetAndUniquifyInstrName(fmha, "fmha-bmm-bmm");
```

Comment:
Nit: why not just use the custom_call_target as a name, and set it at creation time?

Reply:
I think it's cleaner to have this kind of a name rather than using the cutom call target name. The same thing has been done with conv as well. So do u mean having the custom-call name as (say) `__cudnn$fhmaBmmBmm` or `kCudnnfMHABmmBmmCallTarget`?